### PR TITLE
STSMACOM-188, STCOM-505: Create NoteForm component

### PR DESCRIPTION
**Pull request by MaksymDryha**
## Purpose
Different applications may want to have the ability to create/edit notes and attach them to their entities. `<NoteForm />` is a reusable dumb component which apps can use for editing and creating a note. It provides the whole layout for create/edit note pages, including all needed panes, accordions, fields, buttons and the form itself.

## Screens

**Create note page**
image

**Edit note page**
image

Related stories: [STCOM-505](https://issues.folio.org/browse/STCOM-505), [STSMACOM-188](https://issues.folio.org/browse/STSMACOM-188)

## TODO
Discuss where `NavigationModal` component should live. Previously, it was used only by eholdings and was stored in ui-eholdings repository. Now, it's also used by NoteForm and is no longer eholdings-specific. For now, it was temporarily moved to stripes-smart-components, next to all notes-related components

### CODE REVIEW

**Comment by doytch** [comment]
There seems to be some assumptions being made about languages being made [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R47) and [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R80). 

I'm uncertain about whether or not capitalising _always_ makes sense since some languages treat capitalised words differently (German capitalises all nouns). I'm not 100% on it at all, though. Is there any reason why we shouldn't just take the i18n'd string and render it out by default?

It definitely does seem like the LTR default of the second link in `getListItems()` could get us into issues with RTL languages or languages that shift modifiers/counts after the subject in a sentence.

Basically, it seems to me like maybe we should consider supporting (or requiring?) more than just the one translation key.

@JohnC-80 , thoughts?

File: lib/Notes/AssignmentsList/AssignmentsList.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    } = this.props;
+
+    const entityType = (
+      <FormattedMessage
+        id={entityTypeTranslations[type]}
+        values={{ count: 1 }}
+      >
+        {translatedEntityType => (
+          <span className={styles['referred-record__entity-type']}>
+            {capitalize(translatedEntityType)}
```
Comment (Godlevskyi) [inline lib/Notes/AssignmentsList/AssignmentsList.js]: Please remove capitalization and update the localization strings

File: lib/Notes/AssignmentsList/AssignmentsList.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      referredRecord: {
+        name,
+        type,
+      },
+      entityTypeTranslations,
+    } = this.props;
+
+    const entityType = (
+      <FormattedMessage
+        id={entityTypeTranslations[type]}
```
Comment (Godlevskyi) [inline lib/Notes/AssignmentsList/AssignmentsList.js]: should be separate  localization string

File: lib/Notes/AssignmentsList/AssignmentsList.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  getListItems() {
+    const {
+      linkedRecords,
+      entityTypeTranslations,
+    } = this.props;
+
+    return linkedRecords.map(({ count, type }) => (
+      <span className={styles['assignments-list__item']}>
+        {count}
+        &nbsp;
```
Comment (Godlevskyi) [inline lib/Notes/AssignmentsList/AssignmentsList.js]: Please remove it

File: lib/Notes/NoteFields/NoteFields.css [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.editor {
+  min-height: 10em;
```
Comment (Godlevskyi) [inline lib/Notes/NoteFields/NoteFields.css]: Please remove

File: lib/Notes/NoteForm/NoteForm.css [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.note-form-content {
+  margin: 0 auto;
+  max-width: 50rem;
+}
```
Comment (Godlevskyi) [inline lib/Notes/NoteForm/NoteForm.css]: Please remove

File: lib/Notes/NoteForm/NoteForm.js [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+} from '@folio/stripes-components';
+
+import NavigationModal from '../NavigationModal';
+
+import NoteFields from '../NoteFields';
+import AssignmentsList from '../AssignmentsList';
+
+import styles from './NoteForm.css';
+
+const maxTitleLength = 75;
```
Comment (Godlevskyi) [inline lib/Notes/NoteForm/NoteForm.js]: Please rename and move to separate notes/constants.js file

File: lib/Notes/NoteForm/NoteForm.js [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+const maxDetailsLength = 1500;
+
+const invalidCharacters = ['@', '<', '>', '/'];
+
+export default class NoteForm extends Component {
+  static propTypes = {
+    entityTypeTranslations: PropTypes.objectOf(PropTypes.string),
```
Comment (Godlevskyi) [inline lib/Notes/NoteForm/NoteForm.js]: please rename to entityTypeTranslationKeyMap

File: lib/Notes/NoteForm/NoteForm.js [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+        }
+      />
+    );
+  }
+
+  renderCharacterLimitExceededError = (fieldName) => {
+    return <FormattedMessage id={`stripes-smart-components.notes.${fieldName}.lengthLimitExceeded`} />;
+  }
+
+  hasInvalidCharacters = (string) => {
```
Comment (Godlevskyi) [inline lib/Notes/NoteForm/NoteForm.js]: please move to notes/utils.js

**Comment by MaksymDryha** [comment]
> There seems to be some assumptions being made about languages being made [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R47) and [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R80).
> 
> I'm uncertain about whether or not capitalising _always_ makes sense since some languages treat capitalised words differently (German capitalises all nouns). I'm not 100% on it at all, though. Is there any reason why we shouldn't just take the i18n'd string and render it out by default?
> 
> It definitely does seem like the LTR default of the second link in `getListItems()` could get us into issues with RTL languages or languages that shift modifiers/counts after the subject in a sentence.
> 
> Basically, it seems to me like maybe we should consider supporting (or requiring?) more than just the one translation key.
> 
> @JohnC-80 , thoughts?

@doytch, thanks

> There seems to be some assumptions being made about languages being made [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R47) and [here](https://github.com/folio-org/stripes-smart-components/pull/498/files#diff-324aecb3b240be4bb8ba94e06c6432f1R80).
> 
> I'm uncertain about whether or not capitalising _always_ makes sense since some languages treat capitalised words differently (German capitalises all nouns). I'm not 100% on it at all, though. Is there any reason why we shouldn't just take the i18n'd string and render it out by default?
> 
> It definitely does seem like the LTR default of the second link in `getListItems()` could get us into issues with RTL languages or languages that shift modifiers/counts after the subject in a sentence.
> 
> Basically, it seems to me like maybe we should consider supporting (or requiring?) more than just the one translation key.
> 
> @JohnC-80 , thoughts?

Thanks for you comment, @doytch! You are right, it doesn't make sense to use `capitalize()`. It was removed in the last commit

File: lib/Notes/NoteForm/readme.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        value: '1',
+      },
+      {
+        label: 'type 2',
+        value: '2',
+      }
+    ],
+  }), {
+    onSubmit: (values) => {
+      console.log(values);
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/readme.md]: Remove logs.

File: lib/Notes/NoteForm/NoteForm.js [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+                      separator={false}
+                    >
+                      {isEditForm && this.renderNoteMetadata()}
+                      <NoteFields
+                        detailsEditorRef={this.detailsEditorRef}
+                        noteTypes={noteTypes}
+                      />
+                    </Accordion>
+                    <Accordion
+                      onToggle={this.handleSectionToggle}
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/NoteForm.js]: Place callback at the bottom. Check other places.

File: lib/Notes/NoteForm/NoteForm.js [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+                  <Pane
+                    lastMenu={this.renderExpandAllButton()}
+                    defaultWidth="50%"
+                    className={styles['note-form-content']}
+                  >
+                    <Accordion
+                      label={<FormattedMessage id="stripes-smart-components.notes.generalInformation" />}
+                      id="noteForm"
+                      open={noteForm}
+                      onToggle={this.handleSectionToggle}
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/NoteForm.js]: Place callback at the bottom.

File: lib/Notes/NoteForm/NoteForm.js [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    if (title.length > MAX_TITLE_LENGTH) {
+      return this.renderCharacterLimitExceededError('title');
+    }
+
+    return null;
+  }
+
+  renderCharacterLimitExceededError = (fieldName) => {
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/NoteForm.js]: I do not think arrow function is needed here since `this` is not used. Check other places.

File: lib/Notes/NavigationModal/navigation-modal.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    this.setState(INITIAL_MODAL_STATE);
+  };
+
+  onConfirm = () => {
+    this.navigateToNextLocation();
+  };
+
+  navigateToNextLocation() {
+    this.unblock();
+    if ((this.props.historyAction && this.props.historyAction === historyActions.REPLACE)
```
Comment (VictorSoroka1) [inline lib/Notes/NavigationModal/navigation-modal.js]: Make the condition as a variable so it is clear what it stands for.

File: lib/Notes/NavigationModal/navigation-modal.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+} from '@folio/stripes-components';
+
+import { FormattedMessage } from 'react-intl';
+
+const historyActions = {
+  PUSH: 'PUSH',
+  REPLACE: 'REPLACE',
+};
+
+
```
Comment (VictorSoroka1) [inline lib/Notes/NavigationModal/navigation-modal.js]: Remove the excess line.

File: lib/Notes/NoteForm/readme.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+## Description
+Different applications may want to have an ability to create/edit notes and attach them to their entities. `<NoteForm />`
+is a reusable dumb component which apps can use for editing and creating a note. It provides the whole layout for create and edit note pages, including all needed panes, accordions, fields, buttons and the form itself.
+
+## Usage example
+```javascript
+import { NoteForm } from '@folio/stripes-smart-components';
+
+
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/readme.md]: Remove excess line.

File: lib/Notes/NoteForm/NoteForm.js [hunk 16]
```diff
@@ -0,0 +1,288 @@
+import React, { Component, Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { FormattedMessage } from 'react-intl';
+
```
Comment (VictorSoroka1) [inline lib/Notes/NoteForm/NoteForm.js]: Remove excess empty line.

File: lib/Notes/NavigationModal/navigation-modal.js [hunk 17]
```diff
@@ -0,0 +1,109 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import ReactRouterPropTypes from 'react-router-prop-types';
+import { withRouter } from 'react-router';
+import {
```
Comment (VictorSoroka1) [inline lib/Notes/NavigationModal/navigation-modal.js]: Add empty line here before @folio imports.

File: lib/Notes/AssignmentsList/AssignmentsList.js [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+      entityTypeTranslationKeyMap,
+    } = this.props;
+
+    const entityType = <FormattedMessage id={entityTypeTranslationKeyMap[type]} />;
+
+    return (
+      <KeyValue label={entityType}>
+        {name}
+      </KeyValue>
+
```
Comment (VictorSoroka1) [inline lib/Notes/AssignmentsList/AssignmentsList.js]: Empty line.

**Comment by MaksymDryha** [comment]
@VictorSoroka1, thank you for your review! Changes were added with the last commit!

File: lib/Notes/NavigationModal/navigation-modal.js [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  };
+
+  onConfirm = () => {
+    this.navigateToNextLocation();
+  };
+
+  navigateToNextLocation() {
+    this.unblock();
+    const replaceActionRequired =
+      (this.props.historyAction && this.props.historyAction === historyActions.REPLACE)
```
Comment (Yurii-Danylenko) [inline lib/Notes/NavigationModal/navigation-modal.js]: Please do a destructurization of the properties from this.props and this.state
